src/lem/constrainer.rs

```diff
@@ -81,103 +150,571 @@ impl LEM {
+                        if is_concrete_path {
+                            hash_slots.hash3_alloc.push((slots.hash3_idx, i0, i1, i2));
+                            // only once per path
+                        }
+                        // concrete path implies alloc_tgt has the same value as in the current slot
+                        hash_slots.hash3_stacks.implies_stack.push((
+                            concrete_path.clone(),
+                            slots.hash3_idx,
+                            hash,
+                            tag,
+                        ));
+                    }
+                }
+            }
+            AllocHashPreimage::A4(i0, i1, i2, i3) => {
+                match path_kind {
+                    PathKind::Concrete => {
+                        hash_slots
+                            .hash4_stacks
+                            .enforce_stack
+                            .push((slots.hash4_idx, hash, tag));
+                        hash_slots
+                            .hash4_alloc
+                            .push((slots.hash4_idx, i0, i1, i2, i3))
+                    }
+                    PathKind::MaybeConcrete(concrete_path) => {
+                        // if concrete_path is true, push to slots
+                        if is_concrete_path {
+                            hash_slots
+                                .hash4_alloc
+                                .push((slots.hash4_idx, i0, i1, i2, i3));
+                            // only once per path
+                        }
+                        // concrete path implies alloc_tgt has the same value as in the current slot
+                        hash_slots.hash4_stacks.implies_stack.push((
+                            concrete_path.clone(),
+                            slots.hash4_idx,
+                            hash,
+                            tag,
+                        ));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Here we use the implies/enforce logic to contrain tag and hash values.
+    /// We pop stacked elements in order to get virtual path information. When
+    /// many branches are possible we will use the `implies_equal` gadget to
+    /// ensure only concrete paths are implied. Otherwise, only one path exists,
+    /// therefore it must be enforced.
+    /// Hence we use two distinct stacks, one for implications and another one
+    /// for enforcements.
+    /// Beyond that, we need to deal with the possibility of receiving no tag,
+    /// since Unhash operations don't require constraining tags.
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        alloc_ptrs: &HashMap<&String, AllocatedPtr<F>>,
+        implies_stack: &mut Vec<(Boolean, usize, MetaPtr, Option<Tag>)>,
+        enforce_stack: &mut Vec<(usize, MetaPtr, Option<Tag>)>,
+        hash_slots: &HashMap<usize, AllocatedNum<F>>,
```

Comment:
`HashMap<usize, T>` looks a lot like `Vec<T>` to me. Aren't these slots always allocated sequentially? If so, is anything gained by inefficiently simulating a vector with a `HashMap`? If not, what's actually going on that requires allocating out-of-order and/or not having a contiguous range of slot indices?

Reply:
I managed replace it.